src/python/dxpy/sugar/processing.py

```diff
@@ -0,0 +1,508 @@
+    `subprocess` module, and provides access to the results.
+
+    Args:
+        cmds (sequence): Command strings or sequences of command arguments to
+            be chained together.
+        stdout: How to capture stdout of the final process in the chain; can be
+            a string (filename) or file object, in which case the output is written
+            to that file; None, in which case `subprocess.PIPE` is used; True,
+            in which case output is captured to a buffer; or False, in which case
+            stdout is discarded. If a file, any existing file with the same name is
+            overwritten.
+        stderr: How to capture stderr of the final process in the chain (see
+            `stdout` for details).
+        capture_stderr: Whether to capture the contents of stderr of processes
+            other than the final process.
+        echo: Whether to echo commands to the logger before running the commands.
+            Can be overridden by the `run()` method's `echo` parameter.
+        popen_kwargs: Keyword arguments to pass to Popen constructors.
+
+    Todo:
+        Add ability to send input to stdin of first process.
+    """
+    def __init__(
+        self, cmds, stdout=None, stderr=None, capture_stderr=True, echo=None,
+        **popen_kwargs
+    ):
+        self.cmds = cmds
+        self._stdout = stdout
+        self._stdout_type = None
+        self._stderr = stderr
+        self._stderr_type = None
+        self.capture_stderr = capture_stderr
+        self._stderr_buffers = [] if capture_stderr else None
+        self._closed = False
+        self.echo = echo
+        self.popen_kwargs = popen_kwargs
+        self._processes = None
+        self._output_handle = None
+        self._returncode = None
+        # These will hold
+        self.out = None
+        self.err = None
+
+    @property
+    def returncode(self):
+        """
+        The max return code of all the commands in the chain.
+        """
+        if self._returncode is None:
+            returncodes = [
+                proc.returncode for proc in self._processes if proc.poll() is not None
+            ]
+            if returncodes:
+                # TODO: Is setting to max the right behavior?
+                self._returncode = max(returncodes)
+        return self._returncode
+
+    def _init_stdout(self):
+        self._stdout, self._stdout_type, retval = self._init_std(self._stdout)
+        return retval
+
+    def _init_stderr(self):
+        self._stderr, self._stderr_type, retval = self._init_std(self._stderr)
+        return retval
+
+    @staticmethod
+    def _init_std(value):
+        if value is None:
+            return None, _PIPE, subprocess.PIPE
+
+        std_type = _OTHER
+        if value is False:
+            value = None
+        elif value is True:
+            value = Processes._create_temp_outfile()
+            std_type = _BUFFER
+        elif isinstance(value, compat.basestring):
```

Comment:
I believe this is the correct check to perform here. @orodeh any thoughts/comments here? My thought's below:

I need to check if `dx-toolkit` documentation is outdated here. [In the docs](https://github.com/dnanexus/dx-toolkit/tree/a86d952d5142dc01a18d64f50d5049a0f7e39dbe/src/python#python-version-compatibility) we recommend

```python
from dxpy.compat import str
```

Then using `str` from that point onward. 

In this scenario a user is consuming our Class, so:
* in their python2 code they'll send over ASCII `str` (assuming they forget the __future__ imports)
* which would fail a `ininstance(value, str)` in our code since `str` == `unicode`

That's my reasoning why this is the preferred check here. Again @orodeh or @jdidion lmk if this is off.

Reply:
We need to handle both str and unicode in py2, so I think basestring is the right class to check against.